drivers/linkedin.go

```diff
@@ -1,39 +1,72 @@
 package drivers
 
 import (
-	"net/http"
+    "encoding/json"
+    "fmt"
+    "net/http"
+    "os"
 
-	"github.com/danilopolani/gocialite/structs"
-	"golang.org/x/oauth2/linkedin"
+    "github.com/danilopolani/gocialite/structs"
+    "golang.org/x/oauth2/linkedin"
 )
 
 const (
-	linkedinDriverName = "linkedin"
+    linkedinDriverName = "linkedin"
 )
 
 func init() {
-	registerDriver(linkedinDriverName, LinkedInDefaultScopes, LinkedInUserFn, linkedin.Endpoint, LinkedInAPIMap, LinkedInUserMap)
+    registerDriver(linkedinDriverName, LinkedInDefaultScopes, LinkedInUserFn, linkedin.Endpoint, LinkedInAPIMap, LinkedInUserMap)
 }
 
 // LinkedInUserMap is the map to create the User struct
 var LinkedInUserMap = map[string]string{
-	"id":            "ID",
-	"vanityName":    "Username",
-	"firstName":     "FirstName",
-	"lastName":      "LastName",
-	"formattedName": "FullName",
-	"emailAddress":  "Email",
-	"pictureUrl":    "Avatar",
+    "id":                 "ID",
+    "localizedFirstName": "FirstName",
+    "localizedLastName":  "LastName",
 }
 
 // LinkedInAPIMap is the map for API endpoints
 var LinkedInAPIMap = map[string]string{
-	"endpoint":     "https://api.linkedin.com",
-	"userEndpoint": "/v1/people/~:(id,first-name,last-name,formatted-name,email-address,picture-url,maiden-name,headline,location,industry,current-share,num-connections,summary,specialties,positions,public-profile-url)?format=json",
+    "endpoint":      "https://api.linkedin.com",
+    "userEndpoint":  "/v2/me",
+    "emailEndpoint": "/v2/emailAddress?q=members&projection=(elements*(handle~))",
 }
 
 // LinkedInUserFn is a callback to parse additional fields for User
-var LinkedInUserFn = func(client *http.Client, u *structs.User) {}
+var LinkedInUserFn = func(client *http.Client, u *structs.User) {
+    defer func() {
+        if r := recover(); r != nil {
+            fmt.Fprintln(os.Stderr, r)
+        }
+    }()
+    u.FullName = u.FirstName + " " + u.LastName
+    // Retrieve email
+    req, err := client.Get(LinkedInAPIMap["endpoint"] + LinkedInAPIMap["emailEndpoint"])
+    if err != nil {
+        return
+    }
+    defer req.Body.Close()
+    /*
+       {
+         "elements": [
+           {
+             "handle": "urn:li:emailAddress:3775708763",
+             "handle~": {
+               "emailAddress": "hsimpson@linkedin.com"
+             }
+           }
+         ]
+       }
+    */
+    email := make(map[string]interface{})
+    err = json.NewDecoder(req.Body).Decode(&email)
+    if err != nil {
+        return
+    }
+
+    u.Email = email["elements"].([]interface{})[0].(map[string]interface{})["handle~"].(map[string]interface{})["emailAddress"].(string)
+
+}
 
 // LinkedInDefaultScopes contains the default scopes
-var LinkedInDefaultScopes = []string{}
+var LinkedInDefaultScopes = []string{"r_emailaddress", "r_liteprofile", "w_member_social"}
```

Comment:
Why `w_member_social` as a default scope? It's not necessary to retrieve the user details.

Furthermore, it would be better to use `r_basicprofile` to read also the missing fields like `vanityName`

Reply:
We request all the scopes upfront and the developer can optionally change that if required.

Reply:
We should request only the _needed_ scopes to let the package work properly, not the "unnecessary", also because `w_member_social` is a a risky one